python/NexarClient/nexarToken.py

```diff
@@ -0,0 +1,151 @@
+"""Resources for generating Nexar tokens."""
+import os, re, requests
+import base64
+import hashlib
+import http.server
+import webbrowser
+import getpass
+
+from requests_oauthlib import OAuth2Session
+from localService import handlerFactory
+
+HOST_NAME = "localhost"
+PORT = 3000
+REDIRECT_URI = f"http://{HOST_NAME}:{PORT}/login"
+AUTHORITY_URL = "https://identity.nexar.com/connect/authorize"
+PROD_TOKEN_URL = "https://identity.nexar.com/connect/token"
+
+
+def get_token(client_id, client_secret, scopes, refresh_token = None):
+    """Return the Nexar token from the client_id and client_secret provided."""
+
+    if not client_id or not client_secret:
+        raise Exception("client_id and/or client_secret are empty")
+    if not scopes:
+        raise Exception("scope is empty")
+
+    if refresh_token:
+        return get_refresh_token(client_id, client_secret, scopes, refresh_token)
+
+    if (scopes != ["supply.domain"]):
+        return get_token_with_login(client_id, client_secret, scopes)
+
+    token = {}
+    try:
+        token = requests.post(
+            url=PROD_TOKEN_URL,
+            data={
+                "grant_type": "client_credentials",
+                "client_id": client_id,
+                "client_secret": client_secret,
+                "scope": " ".join(scopes)
+            },
+            allow_redirects=False,
+        ).json()
+
+    except Exception:
+        raise
+
+    return token
+
+def get_refresh_token(client_id, client_secret, scopes, refresh_token):
+    """Return the Nexar token from a refresh token."""
+
+    token = {}
+    try:
+        token = requests.post(
+            url=PROD_TOKEN_URL,
+            data={
+                "grant_type": "refresh_token",
+                "refresh_token": refresh_token,
+                "client_id": client_id,
+                "client_secret": client_secret,
+                "scope": " ".join(scopes)
+            },
+        ).json()
+
+    except Exception:
+        raise
+
+    return token
+
+def get_token_with_login(client_id, client_secret, scopes):
+    """Open the Nexar authorization url from the client_id and scope provided."""
+
+    token = {}
+    scope_list = ["openid", "profile", "email"] + scopes
+
+    # PCKE code verifier and challenge
+    code_verifier = base64.urlsafe_b64encode(os.urandom(40)).decode("utf-8")
```

Comment:
I appreciate that this might have been copied over from one of our samples, but we really shouldn't be doing any OAuth2/OIDC flows by hand, and instead use one of the recommended libraries: https://oauth.net/code/python/

Reply:
I think this is a good comment.  I'd say that if there was an example Chung could pick up and put in place here we should do that now.  Otherwise I'm OK if this is a follow-up change.

Reply:
I will take a quick look to see if there exists an example already, if not I can create a ticket in the backlog to amend the OAuth2 flows in the Nexar Developer. This is probably a task best suited to an SE, to ensure best practices are followed - or I'm happy to implement it with the help of an SE.